clampBits() changes

axxel opened the pull request:
Fix most of the remaining core library issues (excluding Test and utilities).

The fix of the `fixBadPixel` copy'n'paste bug (last commit) results in a bunch of rstest hash mismatches.

axxel committed:
Coverity: add missing brake

CID 1298869

axxel committed:
Coverity: fix out-of-bounds-read

CID 1267750. Checked with the dcraw code: the 18 was used for another
set of cameras.

axxel committed:
Coverity: fix off-by-one check in X3fDecoder

CID 1219724, CID 1219725, CID 1294070

taskNo is always < 3. replacing the (wrong) runtime check with a correct
assert.

axxel committed:
Coverity: mark return value as explicitly 'ignored'

CID 981035

axxel committed:
Coverit: fix UNINIT_CTOR in X3fDecoder

CID 1396733

axxel committed:
Coverity: fix UNINIT_CTOR in RawImage

CID 981392, CID 1397424

Also consistently used ` = default` syntax in RawImage class definition.

axxel committed:
Coverity: fix UNINIT_CTROR in DngDecoder

CID 1399558

axxel committed:
Coverity: fix UNINIT_CTOR

CID 1402218

axxel committed:
Coverity: fix UNINIT in RawImageDataU16::fixBadPixel

CID 1025898

axxel committed:
Coverity: fix SIGN_EXTENTION with clampBits

CID 1402217

Improve clampBits interface by fitting it more to the problem at hand:
This function is solely used to clamp some int into an unsigned short
(data type of 16 bit raw image). So have it return a ushort16 and add an
assert(n <= 16). This would effectively allow find a bug if someone
would accidentally clip to 17 bits and then assign the resulting 32-bit
int to a 16bit one.

LebedevRI commented on src/librawspeed/common/Common.h:
```diff
@@ -26,10 +26,7 @@
 #include <vector>           // for vector
-
-#ifndef __BYTE_ORDER__
-#include <cassert> // for assert
```
If you don't know why something is there, maybe not change it

axxel committed:
Coverity: fix copy'n'paste bug in fixBadPixel

CID 1025896, CID 1025897

This 'breaks' a bunch of Panasonic and 2 Leica hashes for obvious
reasons.

LebedevRI commented on src/librawspeed/decoders/CrwDecoder.cpp:
```diff
@@ -193,8 +193,9 @@ void CrwDecoder::decodeMetaDataInternal(CameraMetaData *meta) {
-      int wb_offset = (wb_index < 18) ? "0134567028"[wb_index]-'0' : 0;
-      wb_offset = 1+wb_offset*4;
+      if (wb_index > 9)
+        ThrowIOE("CrwDecoder: invalid white balance index");
```
how is this an input/output issue?

axxel replied:
The type of the exception might be questionable (as are _all_ exception types). The fact is, if `wb_index > 9` then the file is broken.

LebedevRI replied:
True

LebedevRI commented on src/librawspeed/common/Common.h:
```diff
@@ -107,10 +108,11 @@ extern "C" { FILE __iob_func[3] = { *stdin,*stdout,*stderr }; }
-  x = x < 0LL ? 0LL : x;
+// clampBits clamps the given int to the range 0 .. 2^n-1, with n <= 16
+inline ushort16 clampBits(int x, uint32 n) {
+  assert(n <= 16);
```
So what about dngopcodes where it is probably used with wrong argument order?

axxel replied:
I fixed those in #80.

LebedevRI commented:
Picked all but `clampBits()` commit, thanks.
I don't agree with UNINIT_CTOR fixes, they only silence the warning, the real issues are still there.
